### ISSUE TITLE
jax.custom_jvp produces non-deterministically wrong results (exact +2 offset) when called via vmap after processing other custom-pytree types in the same process

### ISSUE THREAD

**Issue by NeilGirdhar** [jax-ml/jax#37389]
### Description

We observe intermittent, non-deterministic incorrect results from a jax.custom_jvp-wrapped function called inside jax.vmap. The failure:                                                                                                                                                                                                                                                                                                                                               
  - Happens in a single sequential process (no threading/multiprocessing)                                                                                                                                                                                                                                                                      
  - Occurs ~30% of runs of the same deterministic code                                                                                                                                                                                                                                                                                         
  - Produces a result that is exactly 2.0 too large every time it fails — not floating-point noise                                                                                                                                                                                                                                             
  - Is triggered by evaluating certain combinations of custom pytree objects earlier in the same process                                                                                                                                                                                                                                       
  - Affects the imaginary part of the log-normalizer under complex inputs, systematically shifting Im(log φ(t)) = η·t to (η+2)·t across all 300 evaluation points                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                                               
The consistent +2 offset (rather than random noise) suggests a JIT cache or compiled kernel selection error rather than a floating-point summation issue.                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                               
Reproducer: Requires the efax library. Full notes [here](https://github.com/NeilGirdhar/efax/blob/cd1781359a32e2c165631338abf77cce6d263d74/.claude/memory/flaky_tests.md) Minimal test sequence:
```fish
  for i in (seq 50)                                         
      if pytest \                                                                                                                                                                                                                                                                                                                              
        "tests/test_ep_from_cf.py::test_ep_from_cf[Bernoulli]" \
        "tests/test_ep_from_cf.py::test_ep_from_cf[Exponential]" \                                                                                                                                                                                                                                                                             
        "tests/test_ep_from_cf.py::test_ep_from_cf[InverseGaussian]" \
        "tests/test_ep_from_cf.py::test_ep_from_cf[UnitVarianceLogNormal]" \                                                                                                                                                                                                                                                                   
        --dist=no -q -p no:randomly                                         
          echo "pass $i"                                                                                                                                                                                                                                                                                                                       
      else                                                  
          echo "FAILED on run $i"; break                                                                                                                                                                                                                                                                                                       
      end                                                   
  end
```

I'm working a better MWE, but would appreciate any ideas about what might be going wrong.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.10.0
jaxlib: 0.10.0
numpy:  2.4.4
python: 3.14.4 (main, Apr  8 2026, 04:02:31) [GCC 15.2.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='Desktop', release='7.0.0-15-generic', version='#15-Ubuntu SMP PREEMPT_DYNAMIC Wed Apr 22 16:06:43 UTC 2026', machine='x86_64')
JAX_PLATFORMS=cpu

$ nvidia-smi
Tue May  5 11:53:31 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.142                Driver Version: 580.142        CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1080        Off |   00000000:01:00.0  On |                  N/A |
|  0%   43C    P5             19W /  180W |    1922MiB /   8192MiB |      2%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            5252      G   /usr/bin/gnome-shell                    600MiB |
|    0   N/A  N/A            5665      G   /usr/bin/Xwayland                        14MiB |
|    0   N/A  N/A            8215    C+G   ...rack-uuid=3190708988185955192        580MiB |
|    0   N/A  N/A           15888      G   ...il/.local/kitty.app/bin/kitty        180MiB |
|    0   N/A  N/A         1821969    C+G   /usr/bin/mpv                            290MiB |
+-----------------------------------------------------------------------------------------+
```

**Comment by NeilGirdhar** [comment]
I think I found the problem: https://github.com/NeilGirdhar/tjax/commit/a99993a502caac3cb4014179b783097315fd577e
